src/progress.rs

```diff
@@ -47,118 +47,187 @@ impl Default for ProgressState {
     }
 }
 
+#[derive(Clone, Debug, Default)]
+struct Configuration {
+    voters: FxHashSet<u64>,
+    learners: FxHashSet<u64>,
+}
+
 /// `ProgressSet` contains several `Progress`es,
 /// which could be `Leader`, `Follower` and `Learner`.
 #[derive(Default, Clone)]
 pub struct ProgressSet {
-    voters: FxHashMap<u64, Progress>,
-    learners: FxHashMap<u64, Progress>,
+    progress: FxHashMap<u64, Progress>,
```

Comment:
How about defining progress as `Vec<Progress>` and `voter`/`learners` as `FxHashMap<u64, u64>`?

Reply:
This is possible.

I actually considered using `FxHashSet<u64>` and `Vec<Progress>` (so a bit farther than this) and just mapping `id = index` in the vector, however we have no guarantee that the ids are not like `1, 1000000`. So that wasn't a good idea.

Your idea here is better. I thought about this too, here's why I didn't do it.

Notably, it would add complexity to:
* `iter()` and `iter_mut()` will need to chain the learners and voters before iterating, and map them.
* `get` and `get_mut` may need to hash twice before getting the index for the Progress. As opposed to once.

It would have the advantage of having a vector lookup for the `Progress`. 

In theory it would:
* Improve `voters()` and `learners()` calls.

You can look at `FxHash` to see that the cost of hashing a `u64` is fairly small:

```rust
fn hash_word(&mut self, word: Self) {
    *self = self.rotate_left(ROTATE).bitxor(word).wrapping_mul($key);
}
```

I think using a `HashMap<u64, Progress>` is also much more clear.

Do you still think we should perform this change? I'd prefer to keep it the way it is.

Reply:
I agree with you.